### ISSUE TITLE
Autoexplore entered an excluded area

### ISSUE THREAD

**Issue by matta** [crawl/crawl#3613]
**PLATFORM and OS or BROWSER:**

webtiles, Firefox 123.0 on Linux

**DCSS VERSION AND GAME INFORMATION:**

https://cbro.berotato.org:8443/#play-dcss-git

0.32-a0-571-gec57dfcd37

**DESCRIPTION:**

1. I autoexplored.
2. Found a sleeping gnoll w/halberd on D:1
3. Excluded the gnoll
4. Explored more.
5. Eventually autoexplore took me back in to the area excluded in step 3.
6. gnoll woke up and shouted

The screenshot is from one step after this happened, because I accidentally moved left one square while filing the bug report. When the gnoll shouted, autoexplore had taken me into the excluded area one square to the right of my current position as pictured:

![Screenshot from 2024-02-29 17-02-01](https://github.com/crawl/crawl/assets/37314/df95ad96-118c-4893-963f-4cb93ce7f370)

**Comment by matthewgranstrom** [comment]
Was the door closed when you placed that exclusion?

**Comment by matta** [comment]
> Was the door closed when you placed that exclusion?

I don't remember, but given that it extends out past the door, and the exclusion doesn't extend out past the door on the east side of the room, maybe the door was open? I had approached the gnoll from the left hallway (NW of where I am in the screenshot).

I'm fine closing this issue with a "don't know what is going on" status.  It is possible I have got something wrong about what happened.

**Comment by matthewgranstrom** [comment]
My point was that exclusions don't extend through doors when placed, so if you could see the door when you excluded the gnoll (maybe even if you couldn't have seen it? unclear to me), the exclusion would have ended there. Then, when you walk into the tile that you walked into, you revealed the tile with the door for the first time since placing the exclusion, and then simultaneously discovered that some monster had opened the door since you placed the exclusion, and that the gnoll behind it had awakened. It's difficult to fix this behavior because to tell the player which doors are open and which are closed out of LOS would reveal tons of information. My usual solution to this is to exclude the monster and also exclude the doors/passageways leading to the monster.

**Comment by matta** [comment]
> My point was that exclusions don't extend through doors when placed

Okay, I think I understand what you're saying. The key observation is that exclusions do extend (become larger) when the player discovers that doors have been opened.

So the theory goes:

1. At the point I placed the exclusion, I could see the south door, so the excluded area did not extend southward past it. This is plausible, since most likely I opened the west door and saw the gnoll (I don't remember).
2. I explored elsewhere, the gnoll awoke (perhaps due to noise, perhaps just after I saw it). It wandered, opened the south door, and wandered precisely back to the center square where I originally saw it.
3. At that very moment, I appeared just beyond the south door. The space autoexplore took me had not previously been excluded because the door was closed. Immediately upon seeing the door opened:

    1.    the exclusion extended southward and covered the square I was in and
    2.    the gnoll saw me and
    3.    coincidentally the gnoll was in the center of the exclusion so it appeared to me as if I had just awoken it.

> My usual solution to this is to exclude the monster and also exclude the doors/passageways leading to the monster.

Good idea!
